File: server/chat_routes.py
import json
import logging
import time
import uuid
from dataclasses import dataclass
from pathlib import Path
from typing import Callable

logger = logging.getLogger("voice.chat")

MAX_UPLOAD = 10 * 1024 * 1024
TEXT_CAP = 50_000
IMAGE_EXTS = {".png", ".jpg", ".jpeg", ".webp", ".gif"}
TEXT_EXTS = {
    ".txt", ".md", ".csv", ".json", ".py", ".js", ".jsx", ".ts", ".tsx",
    ".yaml", ".yml", ".html", ".css", ".sh", ".toml", ".ini", ".log",
}


@dataclass
class ChatContext:
    store: object          # ChatStore
    client: object         # AsyncOpenAI (None in route-only tests)
    ha: object             # HomeAssistant
    system_prompt: str
    default_model: Callable[[], str]
    upload_dir: Path


def register_chat_routes(app, ctx: ChatContext) -> None:
    from fastapi import HTTPException, UploadFile
    from fastapi.responses import Response, StreamingResponse

    from server.chat import run_chat, to_openai_messages

    ctx.upload_dir.mkdir(parents=True, exist_ok=True)

    @app.get("/api/chat/conversations")
    async def list_conversations():
        return ctx.store.list()

    @app.post("/api/chat/conversations")
    async def create_conversation(body: dict | None = None):
        model = (body or {}).get("model") or ctx.default_model()
        return ctx.store.create(model=model)

    @app.get("/api/chat/conversations/{cid}")
    async def get_conversation(cid: str):
        conv = ctx.store.get(cid)
        if conv is None:
            raise HTTPException(404)
        return conv

    @app.patch("/api/chat/conversations/{cid}")
    async def patch_conversation(cid: str, body: dict):
        conv = ctx.store.get(cid)
        if conv is None:
            raise HTTPException(404)
        if "title" in body:
            conv["title"] = str(body["title"])[:80]
        if "model" in body:
            conv["model"] = str(body["model"])
        ctx.store.save(conv)
        return conv

    @app.delete("/api/chat/conversations/{cid}", status_code=204)
    async def delete_conversation(cid: str):
        ctx.store.delete(cid)
        return Response(status_code=204)

    @app.post("/api/chat/upload")
    async def upload(file: UploadFile):
        data = await file.read()
        if len(data) > MAX_UPLOAD:
            raise HTTPException(400, "file too large (max 10 MB)")
        ext = Path(file.filename or "").suffix.lower()
        uid = uuid.uuid4().hex[:12]
        if ext in IMAGE_EXTS:
            (ctx.upload_dir / f"{uid}{ext}").write_bytes(data)
            return {"id": uid, "name": file.filename, "kind": "image"}
        if ext in TEXT_EXTS:
            text = data.decode("utf-8", errors="replace")[:TEXT_CAP]
            (ctx.upload_dir / f"{uid}{ext}").write_bytes(data)
            return {"id": uid, "name": file.filename, "kind": "text", "text": text}
        raise HTTPException(400, f"unsupported file type: {ext or 'unknown'}")

    def sse(obj: dict) -> str:
        return f"data: {json.dumps(obj)}\n\n"

    async def _maybe_title(conv) -> None:
        if conv["title"] != "New chat" or ctx.client is None:
            return
        try:
            resp = await ctx.client.chat.completions.create(
                model=conv["model"],
                messages=[
                    {"role": "user", "content":
                        "Title this conversation in at most 5 words. "
                        "Reply with the title only.\n\nFirst message: "
                        + str(conv["messages"][0].get("content", ""))[:500]},
                ],
            )
            title = (resp.choices[0].message.content or "").strip().strip('"')
            if title:
                conv["title"] = title[:80]
        except Exception:
            logger.debug("auto-title failed", exc_info=True)

    @app.post("/api/chat/conversations/{cid}/messages")
    async def post_message(cid: str, body: dict):
        conv = ctx.store.get(cid)
        if conv is None:
            raise HTTPException(404)
        conv["messages"].append({
            "role": "user",
            "content": body.get("content", ""),
            "attachments": body.get("attachments", []),
        })
        conv["updated"] = time.time()
        ctx.store.save(conv)

        async def stream():
            assistant = {"role": "assistant", "content": "", "thinking": "",
                         "cards": []}
            persisted = False

            def persist():
                nonlocal persisted
                if persisted:
                    return
                persisted = True
                conv["messages"].append(assistant)
                conv["updated"] = time.time()
                ctx.store.save(conv)

            try:
                history = to_openai_messages(conv["messages"], ctx.upload_dir)
                async for event, data in run_chat(
                    ctx.client, conv["model"], ctx.ha, ctx.system_prompt, history
                ):
                    if event == "thinking":
                        assistant["thinking"] += data
                        yield sse({"type": "thinking", "delta": data})
                    elif event == "content":
                        assistant["content"] += data
                        yield sse({"type": "content", "delta": data})
                    elif event == "tool":
                        yield sse({"type": "tool", **data})
                    elif event == "entities":
                        assistant["cards"] = data
                        yield sse({"type": "entities", "entities": data})
                    elif event == "final":
                        assistant["content"] = data["content"]
                        assistant["thinking"] = data["thinking"]
                await _maybe_title(conv)
                persist()
                yield sse({"type": "done", "message": assistant,
                           "title": conv["title"]})
            except Exception as exc:
                logger.exception("chat stream failed")
                persist()
                yield sse({"type": "error", "message": str(exc)})
            finally:
                persist()  # client disconnect mid-stream keeps partial content

        return StreamingResponse(stream(), media_type="text/event-stream")

File: ts-backend/src/chatRoutes.ts
import { readFileSync, readdirSync, writeFileSync } from "node:fs";
import { createChatCompletion } from "./reasoningFallback.js";
import { conversationErrorMessage } from "./conversationError.js";
import { DEFAULT_CONTEXT_WINDOW, type ContextInfo } from "./providers.js";
import { chatCommand, CHAT_HELP, statusReply } from "./chatCommands.js";
import { extname, join } from "node:path";
import type { FastifyInstance } from "fastify";

import { ChatStore, newId, runChat, SAFE_ID, toOpenAiMessages } from "./chat.js";
import { buildChatSystemPrompt, todayLabel } from "./chatPrompt.js";
import type { HomeAssistant } from "./ha.js";
import { toOpenAiTools, type Tool } from "./tools.js";

const MAX_UPLOAD = 10 * 1024 * 1024;
const TEXT_CAP = 50000;
const IMAGE_EXTS = new Set([".png", ".jpg", ".jpeg", ".webp", ".gif"]);
const TEXT_EXTS = new Set([
  ".txt", ".md", ".csv", ".json", ".py", ".js", ".jsx", ".ts", ".tsx",
  ".yaml", ".yml", ".html", ".css", ".sh", ".toml", ".ini", ".log",
]);

const UPLOAD_MEDIA: Record<string, string> = {
  ".png": "image/png", ".jpg": "image/jpeg", ".jpeg": "image/jpeg",
  ".webp": "image/webp", ".gif": "image/gif", ".pdf": "application/pdf",
};

export interface ChatCtx {
  store: ChatStore;
  /** Resolve a (possibly provider-qualified) model id to its client. */
  resolve: (model: string) => { client: any; model: string; providerId?: string; providerName?: string };
  getContextWindow?: (model: string) => Promise<number>;
  getContextInfo?: (model: string) => Promise<ContextInfo>;
  /** Home Assistant tools, offered only when the conversation asks for them. */
  haTools: Tool[];
  /** Shared web tools, including find_in_page; empty without a search key. */
  searchTools: Tool[];
  ha: HomeAssistant | null;
  uploadDir: string;
  /** Assistant name from config. */
  name: string;
  getEntitySummary: () => string;
  getChatInstructions: () => string;
  getDefaultModel: () => string;
}

function findUploadFile(dir: string, uid: string): string | null {
  try {
    for (const e of readdirSync(dir)) {
      if (e.startsWith(uid + ".") && !e.endsWith(".pdftxt")) return e;
    }
  } catch {
    /* dir missing */
  }
  return null;
}

async function maybeTitle(ctx: ChatCtx, conv: any): Promise<void> {
  if (conv.title !== "New chat") return;
  const firstMessage = conv.messages.find((m: any) => m.role === "user" && !m.command);
  if (!firstMessage) return;
  try {
    const { client, model } = ctx.resolve(conv.model);
    const resp = await createChatCompletion<any>(client, {
      model,
      messages: [{
        role: "user",
        content:
          "Title this conversation in at most 5 words. Reply with the title only.\n\nFirst message: " +
          String(firstMessage.content ?? "").slice(0, 500),
      }],
    });
    const title = (resp.choices[0].message.content ?? "").trim().replace(/^"|"$/g, "");
    if (title) conv.title = title.slice(0, 80);
  } catch {
    /* auto-title is best-effort */
  }
}

export function registerChatRoutes(app: FastifyInstance, ctx: ChatCtx): void {
  app.get("/api/chat/conversations", async () => ctx.store.list());

  app.post("/api/chat/conversations", async (req: any) => {
    const body = req.body ?? {};
    const model = body.model || ctx.getDefaultModel();
    return ctx.store.create(model, Boolean(body.homeAssistant), body.webSearch !== false);
  });

  app.get("/api/chat/conversations/:cid", async (req: any, reply: any) => {
    const conv = ctx.store.get(req.params.cid);
    if (!conv) return reply.code(404).send({ detail: "not found" });
    return conv;
  });

  app.patch("/api/chat/conversations/:cid", async (req: any, reply: any) => {
    const conv = ctx.store.get(req.params.cid);
    if (!conv) return reply.code(404).send({ detail: "not found" });
    const body = req.body ?? {};
    if ("title" in body) conv.title = String(body.title).slice(0, 80);
    if ("model" in body) conv.model = String(body.model);
    if ("homeAssistant" in body) conv.homeAssistant = Boolean(body.homeAssistant);
    if ("webSearch" in body) conv.webSearch = Boolean(body.webSearch);
    if ("draftText" in body) {
      const text = String(body.draftText ?? "");
      conv.draftText = text.trim().split(/\s+/u).filter(Boolean).length >= 3 ? text : "";
      conv.updated = Date.now() / 1000;
    }
    ctx.store.save(conv);
    return conv;
  });

  app.delete("/api/chat/conversations/:cid", async (req: any, reply: any) => {
    if (!SAFE_ID.test(req.params.cid)) return reply.code(400).send({ detail: "Invalid conversation ID" });
    try { ctx.store.delete(req.params.cid, ctx.uploadDir); }
    catch { return reply.code(500).send({ detail: "Could not delete the conversation and its files. Please try again." }); }
    return reply.code(204).send();
  });

  app.post("/api/chat/upload", async (req: any, reply: any) => {
    const part = await req.file();
    if (!part) return reply.code(400).send({ detail: "no file" });
    const data: Buffer = await part.toBuffer();
    if (data.length > MAX_UPLOAD) {
      return reply.code(400).send({ detail: "file too large (max 10 MB)" });
    }
    const conversationId = req.query?.conversationId;
    if (conversationId && (!SAFE_ID.test(conversationId) || !ctx.store.get(conversationId))) {
      return reply.code(404).send({ detail: "Conversation no longer exists" });
    }
    const ext = extname(part.filename ?? "").toLowerCase();
    const uid = newId();
    // Register ownership before writing, also covering unsent/removed attachments.
    if (conversationId && (IMAGE_EXTS.has(ext) || TEXT_EXTS.has(ext) || ext === ".pdf")) {
      const conv = ctx.store.get(conversationId);
      conv.uploadIds = [...(conv.uploadIds ?? []), uid];
      ctx.store.save(conv);
    }
    if (IMAGE_EXTS.has(ext)) {
      writeFileSync(join(ctx.uploadDir, uid + ext), data);
      return { id: uid, name: part.filename, kind: "image" };
    }
    if (TEXT_EXTS.has(ext)) {
      const text = data.toString("utf8").slice(0, TEXT_CAP);
      writeFileSync(join(ctx.uploadDir, uid + ext), data);
      return { id: uid, name: part.filename, kind: "text", text };
    }
    if (ext === ".pdf") {
      // NOTE: PDF text extraction (pypdf in the Python backend) is not yet
      // ported to TypeScript; store the file and leave an empty sidecar.
      writeFileSync(join(ctx.uploadDir, uid + ".pdf"), data);
      writeFileSync(join(ctx.uploadDir, uid + ".pdftxt"), "");
      return { id: uid, name: part.filename, kind: "pdf", text: "" };
    }
    return reply.code(400).send({ detail: "unsupported file type: " + (ext || "unknown") });
  });

  app.get("/api/chat/uploads/:uid", async (req: any, reply: any) => {
    const uid = req.params.uid;
    if (!SAFE_ID.test(uid)) return reply.code(404).send();
    const file = findUploadFile(ctx.uploadDir, uid);
    if (!file) return reply.code(404).send();
    const path = join(ctx.uploadDir, file);
    const media = UPLOAD_MEDIA[extname(path).toLowerCase()] ?? "text/plain; charset=utf-8";
    return reply.type(media).send(readFileSync(path));
  });

  app.post("/api/chat/conversations/:cid/messages", async (req: any, reply: any) => {
    const conv = ctx.store.get(req.params.cid);
    if (!conv) return reply.code(404).send({ detail: "not found" });
    const body = req.body ?? {};
    const command = chatCommand(body.content);
    if (!command) {
      conv.draftText = "";
      conv.messages.push({ role: "user", content: body.content ?? "", attachments: body.attachments ?? [] });
      conv.updated = Date.now() / 1000;
      ctx.store.save(conv);
    }

    reply.hijack();
    const raw = reply.raw;
    raw.writeHead(200, {
      "Content-Type": "text/event-stream",
      "Cache-Control": "no-cache",
      Connection: "keep-alive",
    });
    const sse = (obj: any) => raw.write("data: " + JSON.stringify(obj) + "\n\n");

    const assistant: any = { role: "assistant", content: "", thinking: "", cards: [] };
    if (command) assistant.command = command;
    let persisted = false;
    const persist = () => {
      if (command || persisted) return;
      persisted = true;
      // The composer may have saved the next draft while this response streamed.
      const latest = ctx.store.get(conv.id);
      if (!latest) return; // Deleted while the model was responding.
      conv.draftText = latest.draftText ?? "";
      conv.uploadIds = latest.uploadIds ?? [];
      conv.messages.push(assistant);
      conv.updated = Date.now() / 1000;
      ctx.store.save(conv);
    };

    try {
      if (command === "help") {
        assistant.content = CHAT_HELP;
        persist();
        sse({ type: "done", message: assistant, title: conv.title });
        return;
      }
      const pending = command && body.attachments?.length
        ? [{ role: "user", content: "", attachments: body.attachments }] : [];
      const history = toOpenAiMessages([...conv.messages, ...pending], ctx.uploadDir);
      const homeAssistant = Boolean(conv.homeAssistant);
      const searchTools = conv.webSearch !== false ? ctx.searchTools : [];
      const tools = [...searchTools, ...(homeAssistant ? ctx.haTools : [])];
      const system = buildChatSystemPrompt({
        name: ctx.name,
        instructions: ctx.getChatInstructions(),
        today: todayLabel(),
        homeAssistant,
        entitySummary: ctx.getEntitySummary(),
        search: searchTools.length > 0,
      });
      const getCards = homeAssistant && ctx.ha ? (ids: string[]) => ctx.ha!.getCards(ids) : undefined;
      const llm = ctx.resolve(conv.model);
      if (command === "status") {
        const context = await ctx.getContextInfo?.(conv.model)
          ?? { tokens: DEFAULT_CONTEXT_WINDOW, source: "default" as const };
        Object.assign(assistant, statusReply(llm.model, llm.providerName ?? llm.providerId ?? "Unknown provider", context,
          [{ role: "system", content: system }, ...history], toOpenAiTools(tools)));
        persist();
        sse({ type: "done", message: assistant, title: conv.title });
        return;
      }
      const contextWindow = await ctx.getContextWindow?.(conv.model) ?? DEFAULT_CONTEXT_WINDOW;
      sse({ type: "debug", event: "context", data: {
        model: llm.model, providerId: llm.providerId, providerName: llm.providerName,
        contextWindow,
      } });
      for await (const [event, data] of runChat(
        llm.client, llm.model, tools, system, history, getCards,
      )) {
        if (event === "thinking") {
          assistant.thinking += data;
          sse({ type: "thinking", delta: data });
        } else if (event === "content") {
          assistant.content += data;
          sse({ type: "content", delta: data });
        } else if (event === "tool") {
          sse({ type: "tool", ...data });
        } else if (event === "search") {
          assistant.search = data;
          sse({ type: "search", ...data });
        } else if (event === "debug") {
          sse({ type: "debug", ...data });
        } else if (event === "entities") {
          assistant.cards = data;
          sse({ type: "entities", entities: data });
        } else if (event === "final") {
          assistant.content = data.content;
          assistant.thinking = data.thinking;
        }
      }
      await maybeTitle(ctx, conv);
      persist();
      sse({ type: "done", message: assistant, title: conv.title });
    } catch (e: any) {
      if (assistant.search?.phase === "start") assistant.search.phase = "error";
      persist();
      sse({ type: "error", message: conversationErrorMessage(e) });
    } finally {
      persist();
      raw.end();
    }
  });
}
